File: benchmarks/crown/buffer/Output/tmp/test/test-buffer.c

//
// test.c
//
// Copyright (c) 2012 TJ Holowaychuk <tj@vision-media.ca>
//

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

//
// buffer.h
//
// Copyright (c) 2012 TJ Holowaychuk <tj@vision-media.ca>
//

#ifndef BUFFER_H
#define BUFFER_H 1

#include <sys/types.h>

/*
 * Default buffer size.
 */

#ifndef BUFFER_DEFAULT_SIZE
#define BUFFER_DEFAULT_SIZE 64
#endif

/*
 * Buffer struct.
 */

typedef struct {
  size_t len;
  char *alloc;
  char *data;
} buffer_t;

// prototypes

buffer_t *
buffer_new();

buffer_t *
buffer_new_with_size(size_t n);

buffer_t *
buffer_new_with_string(char *str);

buffer_t *
buffer_new_with_string_length(char *str, size_t len);

buffer_t *
buffer_new_with_copy(char *str);

size_t
buffer_size(buffer_t *self);

size_t
buffer_length(buffer_t *self);

void
buffer_free(buffer_t *self);

int
buffer_prepend(buffer_t *self, char *str);

int
buffer_append(buffer_t *self, const char *str);

int
buffer_appendf(buffer_t *self, const char *format, ...);

int
buffer_append_n(buffer_t *self, const char *str, size_t len);

int
buffer_equals(buffer_t *self, buffer_t *other);

ssize_t
buffer_indexof(buffer_t *self, char *str);

buffer_t *
buffer_slice(buffer_t *self, size_t from, ssize_t to);

ssize_t
buffer_compact(buffer_t *self);

void
buffer_fill(buffer_t *self, int c);

void
buffer_clear(buffer_t *self);

void
buffer_trim_left(buffer_t *self);

void
buffer_trim_right(buffer_t *self);

void
buffer_trim(buffer_t *self);

void
buffer_print(buffer_t *self);

#define buffer_string(self) (self->data)

#endif

void
equal(char *a, char *b) {
  if (strcmp(a, b)) {
    printf("\n");
    printf("  expected: '%s'\n", a);
    printf("    actual: '%s'\n", b);
    printf("\n");
    exit(1);
  }
}

void
test_buffer_new() {
  buffer_t *buf = buffer_new();
  assert(BUFFER_DEFAULT_SIZE == buffer_size(buf));
  assert(0 == buffer_length(buf));
  buffer_free(buf);
}

void
test_buffer_new_with_size() {
  buffer_t *buf = buffer_new_with_size(1024);
  assert(1024 == buffer_size(buf));
  assert(0 == buffer_length(buf));
  buffer_free(buf);
}

void
test_buffer_append() {
  buffer_t *buf = buffer_new();
  assert(0 == buffer_append(buf, "Hello"));
  assert(0 == buffer_append(buf, " World"));
  assert(strlen("Hello World") == buffer_length(buf));
  equal("Hello World", buffer_string(buf));
  buffer_free(buf);
}

void
test_buffer_append_n() {
  buffer_t *buf = buffer_new();
  assert(0 == buffer_append_n(buf, "subway", 3));
  assert(0 == buffer_append_n(buf, "marines", 6));
  assert(strlen("submarine") == buffer_length(buf));
  equal("submarine", buffer_string(buf));
  buffer_free(buf);
}

void
test_buffer_append__grow() {
  buffer_t *buf = buffer_new_with_size(10);
  assert(0 == buffer_append(buf, "Hello"));
  assert(0 == buffer_append(buf, " tobi"));
  assert(0 == buffer_append(buf, " was"));
  assert(0 == buffer_append(buf, " here"));

  char *str = "Hello tobi was here";
  equal(str, buffer_string(buf));
  assert(1024 == buffer_size(buf));
  assert(strlen(str) == buffer_length(buf));
  buffer_free(buf);
}

void
test_buffer_prepend() {
  buffer_t *buf = buffer_new();
  assert(0 == buffer_append(buf, " World"));
  assert(0 == buffer_prepend(buf, "Hello"));
  assert(strlen("Hello World") == buffer_length(buf));
  equal("Hello World", buffer_string(buf));
  buffer_free(buf);
}

void
test_buffer_slice() {
  buffer_t *buf = buffer_new();
  buffer_append(buf, "Tobi Ferret");

  buffer_t *a = buffer_slice(buf, 2, 8);
  equal("Tobi Ferret", buffer_string(buf));
  equal("bi Fer", buffer_string(a));

  buffer_free(buf);
  buffer_free(a);
}

void
test_buffer_slice__range_error() {
  buffer_t *buf = buffer_new_with_copy("Tobi Ferret");
  buffer_t *a = buffer_slice(buf, 10, 2);
  assert(NULL == a);
  buffer_free(buf);
}

void
test_buffer_slice__end() {
  buffer_t *buf = buffer_new_with_copy("Tobi Ferret");

  buffer_t *a = buffer_slice(buf, 5, -1);
  equal("Tobi Ferret", buffer_string(buf));
  equal("Ferret", buffer_string(a));

  buffer_t *b = buffer_slice(buf, 5, -3);
  equal("Ferr", buffer_string(b));

  buffer_t *c = buffer_slice(buf, 8, -1);
  equal("ret", buffer_string(c));

  buffer_free(buf);
  buffer_free(a);
  buffer_free(b);
  buffer_free(c);
}

void
test_buffer_slice__end_overflow() {
  buffer_t *buf = buffer_new_with_copy("Tobi Ferret");
  buffer_t *a = buffer_slice(buf, 5, 1000);
  equal("Tobi Ferret", buffer_string(buf));
  equal("Ferret", buffer_string(a));
  buffer_free(a);
  buffer_free(buf);
}

void
test_buffer_equals() {
  buffer_t *a = buffer_new_with_copy("Hello");
  buffer_t *b = buffer_new_with_copy("Hello");

  assert(1 == buffer_equals(a, b));

  buffer_append(b, " World");
  assert(0 == buffer_equals(a, b));

  buffer_free(a);
  buffer_free(b);
}

void test_buffer_formatting() {
  buffer_t *buf = buffer_new();
  int result = buffer_appendf(buf, "%d %s", 3, "cow");
  assert(0 == result);
  equal("3 cow", buffer_string(buf));
  result = buffer_appendf(buf, " - 0x%08X", 0xdeadbeef);
  assert(0 == result);
  equal("3 cow - 0xDEADBEEF", buffer_string(buf));
  buffer_free(buf);
}

void
test_buffer_indexof() {
  buffer_t *buf = buffer_new_with_copy("Tobi is a ferret");

  ssize_t i = buffer_indexof(buf, "is");
  assert(5 == i);

  i = buffer_indexof(buf, "a");
  assert(8 == i);

  i = buffer_indexof(buf, "something");
  assert(-1 == i);

  buffer_free(buf);
}

void
test_buffer_fill() {
  buffer_t *buf = buffer_new_with_copy("Hello");
  assert(5 == buffer_length(buf));

  buffer_fill(buf, 0);
  assert(0 == buffer_length(buf));
  buffer_free(buf);
}

void
test_buffer_clear() {
  buffer_t *buf = buffer_new_with_copy("Hello");
  assert(5 == buffer_length(buf));

  buffer_clear(buf);
  assert(0 == buffer_length(buf));
  buffer_free(buf);
}

void
test_buffer_trim() {
  buffer_t *buf = buffer_new_with_copy("  Hello\n\n ");
  buffer_trim(buf);
  equal("Hello", buffer_string(buf));
  buffer_free(buf);

  buf = buffer_new_with_copy("  Hello\n\n ");
  buffer_trim_left(buf);
  equal("Hello\n\n ", buffer_string(buf));
  buffer_free(buf);

  buf = buffer_new_with_copy("  Hello\n\n ");
  buffer_trim_right(buf);
  equal("  Hello", buffer_string(buf));
  buffer_free(buf);
}

void
test_buffer_compact() {
  buffer_t *buf = buffer_new_with_copy("  Hello\n\n ");
  buffer_trim(buf);
  assert(5 == buffer_length(buf));
  assert(10 == buffer_size(buf));

  ssize_t removed = buffer_compact(buf);
  assert(5 == removed);
  assert(5 == buffer_length(buf));
  assert(5 == buffer_size(buf));
  equal("Hello", buffer_string(buf));

  buffer_free(buf);
}

void
test_buffer_prepend_issue_15() {
  buffer_t *file = buffer_new();
  assert(0 == buffer_append(file, "layout.bk.html"));
  assert(0 == buffer_prepend(file, "./example/"));
  assert(strlen("./example/layout.bk.html") == buffer_length(file));
  equal("./example/layout.bk.html", buffer_string(file));
  buffer_free(file);
}

int
main(){
  test_buffer_new();
  test_buffer_new_with_size();
  test_buffer_append();
  test_buffer_append__grow();
  test_buffer_append_n();
  test_buffer_prepend();
  test_buffer_slice();
  test_buffer_slice__range_error();
  test_buffer_slice__end();
  test_buffer_slice__end_overflow();
  test_buffer_equals();
  test_buffer_formatting();
  test_buffer_indexof();
  test_buffer_fill();
  test_buffer_clear();
  test_buffer_trim();
  test_buffer_compact();
  test_buffer_prepend_issue_15();
  printf("\n  \e[32m\u2713 \e[90mok\e[0m\n\n");
  return 0;
}

File: benchmarks/crown/buffer/Output/Output/test_project/tests/test_buffer.rs
use ntest::timeout;
use test_project::buffer::{
    buffer_append, buffer_append_n, buffer_appendf, buffer_clear, buffer_compact, buffer_equals,
    buffer_fill, buffer_free, buffer_indexof, buffer_length, buffer_new, buffer_new_with_copy,
    buffer_new_with_size, buffer_prepend, buffer_size, buffer_slice, buffer_trim, buffer_trim_left,
    buffer_trim_right, Buffer, BUFFER_DEFAULT_SIZE,
};
#[test]
#[timeout(60000)]
pub fn test_buffer_prepend() {
    let mut buf = buffer_new().unwrap();
    assert_eq!(0, buffer_append(&mut buf, " World"));
    assert_eq!(0, buffer_prepend(&mut buf, "Hello".to_string()));
    assert_eq!("Hello World".len(), buffer_length(&buf));
    equal("Hello World", &buf.data);
    buffer_free(buf);
}

pub fn equal(a: &str, b: &str) {
    if a != b {
        println!();
        println!("  expected: '{}'", a);
        println!("    actual: '{}'", b);
        println!();
        std::process::exit(1);
    }
}

#[test]
#[timeout(60000)]
pub fn test_buffer_new() {
    let buf = buffer_new().unwrap();
    assert_eq!(BUFFER_DEFAULT_SIZE, buffer_size(&buf));
    assert_eq!(0, buffer_length(&buf));
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_new_with_size() {
    let buf = buffer_new_with_size(1024).unwrap();
    assert_eq!(1024, buffer_size(&buf));
    assert_eq!(0, buffer_length(&buf));
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_slice__end_overflow() {
    let buf = buffer_new_with_copy("Tobi Ferret".to_string()).unwrap();
    let a = buffer_slice(&buf, 5, 1000).unwrap();
    equal("Tobi Ferret", &buf.data);
    equal("Ferret", &a.data);
    buffer_free(a);
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_compact() {
    let mut buf = buffer_new_with_copy("  Hello\n\n ".to_string()).unwrap();
    buffer_trim(&mut buf);
    assert_eq!(5, buffer_length(&buf));
    assert_eq!(10, buffer_size(&buf));

    let removed = buffer_compact(&mut buf);
    assert_eq!(5, removed);
    assert_eq!(5, buffer_length(&buf));
    assert_eq!(5, buffer_size(&buf));
    equal("Hello", &buf.data);

    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_append() {
    let mut buf = buffer_new().unwrap();
    assert_eq!(0, buffer_append(&mut buf, "Hello"));
    assert_eq!(0, buffer_append(&mut buf, " World"));
    assert_eq!("Hello World".len(), buffer_length(&buf));
    equal("Hello World", &buf.data);
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_append_n() {
    let mut buf = buffer_new().unwrap();
    assert_eq!(0, buffer_append_n(&mut buf, "subway", 3));
    assert_eq!(0, buffer_append_n(&mut buf, "marines", 6));
    assert_eq!("submarine".len(), buffer_length(&buf));
    equal("submarine", &buf.data);
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_append__grow() {
    let mut buf = buffer_new_with_size(10).unwrap();
    assert_eq!(0, buffer_append(&mut buf, "Hello"));
    assert_eq!(0, buffer_append(&mut buf, " tobi"));
    assert_eq!(0, buffer_append(&mut buf, " was"));
    assert_eq!(0, buffer_append(&mut buf, " here"));

    let str = "Hello tobi was here";
    equal(str, &buf.data);
    assert_eq!(1024, buffer_size(&buf));
    assert_eq!(str.len(), buffer_length(&buf));
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_prepend_issue_15() {
    let mut file = buffer_new().unwrap();
    assert_eq!(0, buffer_append(&mut file, "layout.bk.html"));
    assert_eq!(0, buffer_prepend(&mut file, "./example/".to_string()));
    assert_eq!("./example/layout.bk.html".len(), buffer_length(&file));
    equal("./example/layout.bk.html", &file.data);
    buffer_free(file);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_fill() {
    let mut buf = buffer_new_with_copy("Hello".to_string()).unwrap();
    assert_eq!(5, buffer_length(&buf));

    buffer_fill(&mut buf, 0);
    assert_eq!(0, buffer_length(&buf));
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_indexof() {
    let buf = buffer_new_with_copy(String::from("Tobi is a ferret")).unwrap();

    let i = buffer_indexof(&buf, String::from("is"));
    assert_eq!(5, i);

    let i = buffer_indexof(&buf, String::from("a"));
    assert_eq!(8, i);

    let i = buffer_indexof(&buf, String::from("something"));
    assert_eq!(-1, i);

    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_trim() {
    let mut buf = buffer_new_with_copy("  Hello\n\n ".to_string()).unwrap();
    buffer_trim(&mut buf);
    equal("Hello", &buf.data);

    buffer_free(buf);

    let mut buf = buffer_new_with_copy("  Hello\n\n ".to_string()).unwrap();
    buffer_trim_left(&mut buf);
    equal("Hello\n\n ", &buf.data);

    buffer_free(buf);

    let mut buf = buffer_new_with_copy("  Hello\n\n ".to_string()).unwrap();
    buffer_trim_right(&mut buf);
    equal("  Hello", &buf.data);

    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_slice__range_error() {
    let buf = buffer_new_with_copy("Tobi Ferret".to_string()).unwrap();
    let a = buffer_slice(&buf, 10, 2);
    assert!(a.is_err());
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_slice__end() {
    let buf = buffer_new_with_copy("Tobi Ferret".to_string()).unwrap();

    let a = buffer_slice(&buf, 5, -1).unwrap();
    equal("Tobi Ferret", &buf.data);
    equal("Ferret", &a.data);

    let b = buffer_slice(&buf, 5, -3).unwrap();
    equal("Ferr", &b.data);

    let c = buffer_slice(&buf, 8, -1).unwrap();
    equal("ret", &c.data);

    buffer_free(buf);
    buffer_free(a);
    buffer_free(b);
    buffer_free(c);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_clear() {
    let mut buf = buffer_new_with_copy("Hello".to_string()).unwrap();
    assert_eq!(5, buffer_length(&buf));

    buffer_clear(&mut buf);
    assert_eq!(0, buffer_length(&buf));
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_formatting() {
    let mut buf = buffer_new().expect("Failed to create buffer");
    let result = buffer_appendf(
        &mut buf,
        "%d %s".to_string(),
        vec!["3".to_string(), "cow".to_string()],
    );
    assert_eq!(0, result);
    equal("3 cow", &buf.data);
    let result = buffer_appendf(
        &mut buf,
        " - 0x%08X".to_string(),
        vec!["0xDEADBEEF".to_string()],
    );
    assert_eq!(0, result);
    equal("3 cow - 0xDEADBEEF", &buf.data);
    buffer_free(buf);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_equals() {
    let a = buffer_new_with_copy("Hello".to_string()).unwrap();
    let b = buffer_new_with_copy("Hello".to_string()).unwrap();

    assert!(buffer_equals(&a, &b));

    buffer_append(&mut b.clone(), " World");
    assert!(!buffer_equals(&a, &b));

    buffer_free(a);
    buffer_free(b);
}

#[test]
#[timeout(60000)]
pub fn test_buffer_slice() {
    let mut buf = buffer_new().unwrap();
    buffer_append(&mut buf, "Tobi Ferret");

    let a = buffer_slice(&buf, 2, 8).unwrap();
    equal("Tobi Ferret", &buf.data);
    equal("bi Fer", &a.data);

    buffer_free(buf);
    buffer_free(a);
}

fn nearest_multiple_of(base: usize, n: usize) -> usize {
    ((n + base - 1) / base) * base
}
